cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -505,29 +636,31 @@ class InputState {
         time_type_id_(schema_->fields()[time_col_index_]->type()->id()),
         key_type_id_(key_col_index.size()),
         key_hasher_(key_hasher),
+        node_(node),
+        index_(index),
         must_hash_(must_hash),
         may_rehash_(may_rehash),
         tolerance_(tolerance),
-        memo_(/*no_future=*/index == 0 || !tolerance.positive) {
+        memo_(node, index, /*no_future=*/index == 0 || !tolerance.positive) {
     for (size_t k = 0; k < key_col_index_.size(); k++) {
       key_type_id_[k] = schema_->fields()[key_col_index_[k]]->type()->id();
     }
   }
 
   static Result<std::unique_ptr<InputState>> Make(
       size_t index, TolType tolerance, bool must_hash, bool may_rehash,
-      KeyHasher* key_hasher, ExecNode* node, ExecNode* output,
+      KeyHasher* key_hasher, AsofJoinNode& node, ExecNode* output,
       std::atomic<int32_t>& backpressure_counter,
       const std::shared_ptr<arrow::Schema>& schema, const col_index_t time_col_index,
       const std::vector<col_index_t>& key_col_index) {
     constexpr size_t low_threshold = 4, high_threshold = 8;
     std::unique_ptr<BackpressureControl> backpressure_control =
-        std::make_unique<BackpressureController>(node, output, backpressure_counter);
+        std::make_unique<BackpressureController>(&node, output, backpressure_counter);
```

Comment:
Why change from pointer to reference?

Reply:
IIRC, this just signals `node` can be modified.

Reply:
Seem - I don't see other places that using a reference to indicate that a object can be modified. Is this done in other places?